### PR TITLE
Use thread-safe localtime function in unix distributions [15337]

### PR DESCRIPTION
**Pull request by rsanchez15**
From `localtime` [cplusplus.com](https://cplusplus.com/reference/ctime/localtime/) documentation:

"_The function accesses the object pointed by timer.
The function also accesses and modifies a shared internal object, which may introduce data races on concurrent calls to [gmtime](https://cplusplus.com/gmtime) and localtime. Some libraries provide an alternative function that avoids this data race: localtime_r (non-portable)._"

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [ ] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [ ] Documentation builds and tests pass locally. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- [ ] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/fastdds/log/Log.cpp [hunk 1]
```diff
@@ -331,13 +331,16 @@ void Log::get_timestamp(
 #if defined(_WIN32)
     struct tm timeinfo;
     localtime_s(&timeinfo, &now_c);
-    stream << std::put_time(&timeinfo, "%F %T") << "." << std::setw(3) << std::setfill('0') << ms << " ";
     //#elif defined(__clang__) && !defined(std::put_time) // TODO arm64 doesn't seem to support std::put_time
     //    (void)now_c;
     //    (void)ms;
+#elif defined(__unix__)
```
Comment (jsantiago-eProsima) [inline src/cpp/fastdds/log/Log.cpp:337]: According to ctime(3) manpage there's an additional define that should be checked:

`Feature Test Macro Requirements for glibc (see feature_test_macros(7)):

       asctime_r(), ctime_r(), gmtime_r(), localtime_r():
              _POSIX_C_SOURCE
                  || /* Glibc versions <= 2.19: */ _BSD_SOURCE || _SVID_SOURCE`

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
It would be great to have a single method in an internal utility class/namespace for getting a string with the current time, and use it from both sources.

File: src/cpp/rtps/transport/shared_mem/SharedMemLog.hpp [hunk 2]
```diff
@@ -327,13 +327,16 @@ class PacketsLog
     #if defined(_WIN32)
         struct tm timeinfo;
         localtime_s(&timeinfo, &now_c);
-        stream << std::put_time(&timeinfo, "%T") << "." << std::setw(3) << std::setfill('0') << ms << " ";
         //#elif defined(__clang__) && !defined(std::put_time) // TODO arm64 doesn't seem to support std::put_time
         //    (void)now_c;
         //    (void)ms;
+    #elif defined(__unix__)
+        std::tm timeinfo;
```
Comment (jsantiago-eProsima) [inline src/cpp/rtps/transport/shared_mem/SharedMemLog.hpp:334]: Same as the previous comment.

**Review by jsantiago-eProsima — CHANGES_REQUESTED** [review]

File: src/cpp/utils/SystemInfo.hpp [hunk 3]
```diff
@@ -82,6 +82,15 @@ class SystemInfo
...
         static SystemInfo singleton;
+
+        // From ctime(3) linux man page:
+        // According to POSIX.1-2004, localtime() is required to behave as though tzset(3) was called, while
+        // localtime_r() does not have this requirement. For portable code tzset(3) should be called before
+        // localtime_r().
+#if (_POSIX_C_SOURCE >= 1) || defined(_XOPEN_SOURCE) || defined(_POSIX_SOURCE)
+        tzset();
+#endif // if (_POSIX_C_SOURCE >= 1) || defined(_XOPEN_SOURCE) || defined(_POSIX_SOURCE)
+
```
Comment (MiguelCompany) [inline src/cpp/utils/SystemInfo.hpp:93]: I think it is better to have this code in the constructor of the class.

The conditional should also match the one on `SystemInfo::get_timestamp`

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/utils/SystemInfo.hpp (hunk 3 above)
Comment (rsanchez15) [reply src/cpp/utils/SystemInfo.hpp]: Addressed in 43e8b4f

File: src/cpp/fastdds/log/Log.cpp (hunk 1 above)
Comment (rsanchez15) [reply src/cpp/fastdds/log/Log.cpp]: Addressed in 66c46be

File: src/cpp/rtps/transport/shared_mem/SharedMemLog.hpp (hunk 2 above)
Comment (rsanchez15) [reply src/cpp/rtps/transport/shared_mem/SharedMemLog.hpp]: Same as the previous answer.

**Review by MiguelCompany — APPROVED** [review]
LGTM with green CI

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by rsanchez15** [comment]
@richiprosima Please test this

**Review by MiguelCompany — APPROVED** [review]

**Comment by rsanchez15** [comment]
@richiprosima Please test this

**Review by MiguelCompany — APPROVED** [review]
LGTM with green CI

**Pull request merged**